qcodes_contrib_drivers/drivers/ERAInstruments/erasynth.py

```diff
@@ -0,0 +1,900 @@
+
+        return response
+
+    def ask_raw(self, cmd: str) -> str:
+        """
+        Detects special commands for which the `get_configuration` will be used.
+
+        This makes it convenient to not implement individual getters for most commands.
+        """
+        if cmd[1:].startswith("RA:"):
+            response = self.get_configuration(cmd[1 + len("RA:") :])
+        elif cmd[1:].startswith("RD:"):
+            response = self.get_diagnostic_status(cmd[1 + len("RD:") :])
+        else:
+            response = super().ask_raw(cmd)
+
+        return response
+
+    def write(self, cmd: str) -> None:
+        """Writes a command to the communication channel of the instrument.
+
+        Commands are prefixed with `">"` as required by the ERASynth.
+
+        NB the read buffer is discarded before and after reading one line.
+        """
+        self.clear_read_buffer()
+        super().write(f">{cmd}")
+        self.clear_read_buffer()
+
+    def write_raw(self, cmd: str) -> None:
+        """
+        For some commands we confirm that the value has been set correctly.
+
+        This is only possible for configurations that can be retrieved from the
+        instrument.
+        """
+        is_readable_cmd = False
+        for command in _CMD_TO_JSON_MAPPING:
+            if cmd[1:].startswith(command):
+                is_readable_cmd = True
+                break
+
+        if is_readable_cmd:
+            json_key = _CMD_TO_JSON_MAPPING[command]
+            cmd_arg = cmd[1 + len(command) :]
+            while True:
+                super().write_raw(cmd)
+                self.clear_read_buffer()
+                if self.get_configuration(json_key) == cmd_arg:
+                    break
+        else:
+            super().write_raw(cmd)
+
+    def _get_json(self, cmd: str, first_key: str) -> str:
+        """
+        Sends command and reads result until the result looks like a JSON.
+        """
+        timeout = 10
+        t_start = time.time()
+        first_key = '{"' + first_key + '"'
+        while True:
+            read_line = self.ask(cmd)
+            # Ensure the line contains the json
+            if first_key in read_line and read_line[-1] == "}":
+                break
+            if time.time() > t_start + timeout:
+                raise TimeoutError(
+                    f"Failed to query JSON within {timeout} s. "
+                    f"Command {cmd!r} failed."
+                )
+
+        return "".join(["{", *read_line.split("{")[1:]])
+
+    # ERASynth specific methods
+
+    def get_configuration(self, par_name: str = None) -> Union[Dict[str, str], str]:
+        """
+        Returns the configuration JSON that contains all parameters.
+        """
+        config_json = json.loads(self._get_json("RA", "rfoutput"))
+
+        return config_json if par_name is None else config_json[par_name]
+
+    def get_diagnostic_status(self, par_name: str = None) -> Union[Dict[str, str], str]:
```

Comment:
Shouldn't the output be always a `str`, then?

Reply:
no, it can be called manually with no arguments to get the JSON that the instrument returns for more "manual debugging"

Reply:
PS if mypy is still not happy in the `ask_raw`, feel free to cast `str(self. get_diagnostic_status(...))`